File: moyai_app/lib/explore/model/entities/venue.dart
class Venue {
  final String venueId;
  final String name;
  final String phone;
  final String? signature;
  final String email;
  final List<String> openingHours;
  final String frontImageOfBusinessDescribe;
  final bool consentObtained;
  final String? frontImage;
  final List<String> accessibleParkingAvailable;
  final List<String> closestAccessibleParking;
  final List<String> distanceToClosestAccessibleParking;
  final String assessorName;
  final String businessName;
  final String businessType;
  final List<String> address;
  final String frontImageOfBusiness;
  final Coordinates? coordinates;
  final String createdAt;
  final String updatedAt;

  Venue({
    required this.venueId,
    required this.name,
    required this.phone,
    this.signature,
    required this.email,
    required this.openingHours,
    required this.frontImageOfBusinessDescribe,
    required this.consentObtained,
    this.frontImage,
    required this.accessibleParkingAvailable,
    required this.closestAccessibleParking,
    required this.distanceToClosestAccessibleParking,
    required this.assessorName,
    required this.businessName,
    required this.businessType,
    required this.address,
    required this.frontImageOfBusiness,
    required this.coordinates,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Venue.fromJson(Map<String, dynamic> json) {
    return Venue(
      venueId: json['venueId'],
      name: json['name'],
      phone: json['phone'],
      signature: json['signature'],
      email: json['email'],
      openingHours: List<String>.from(json['openingHours']),
      frontImageOfBusinessDescribe: json['frontImageOfBusinessDescribe'],
      consentObtained: json['consentObtained'],
      frontImage: json['frontImage'],
      accessibleParkingAvailable: List<String>.from(json['accessibleParkingAvailable']),
      closestAccessibleParking: List<String>.from(json['closestAccessibleParking']
          .replaceAll(RegExp(r'[{}]'), '')
          .split(',')),
      distanceToClosestAccessibleParking: List<String>.from(json['distanceToClosestAccessibleParking']
          .replaceAll(RegExp(r'[{}]'), '')
          .split(',')),
      assessorName: json['assessorName'],
      businessName: json['businessName'],
      businessType: json['businessType'],
      address: List<String>.from(json['address']),
      frontImageOfBusiness: json['frontImageOfBusiness'],
      coordinates: Coordinates.fromJson(json['coordinates']),
      createdAt: json['createdAt'],
      updatedAt: json['updatedAt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'venueId': venueId,
      'name': name,
      'phone': phone,
      'signature': signature,
      'email': email,
      'openingHours': openingHours,
      'frontImageOfBusinessDescribe': frontImageOfBusinessDescribe,
      'consentObtained': consentObtained,
      'frontImage': frontImage,
      'accessibleParkingAvailable': accessibleParkingAvailable,
      'closestAccessibleParking': closestAccessibleParking,
      'distanceToClosestAccessibleParking': distanceToClosestAccessibleParking,
      'assessorName': assessorName,
      'businessName': businessName,
      'businessType': businessType,
      'address': address,
      'frontImageOfBusiness': frontImageOfBusiness,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
    };
  }
}


class Coordinates {
  final double latitude;
  final double longitude;

  Coordinates({
    required this.latitude,
    required this.longitude,
  });

  factory Coordinates.fromJson(Map<String, dynamic> json) {
    return Coordinates(
      latitude: json['latitude'],
      longitude: json['longitude'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'latitude': latitude,
      'longitude': longitude,
    };
  }
}

File: backend-code/src/db/entities/Venue.ts
import {
  Entity,
  BaseEntity,
  Column,
  PrimaryGeneratedColumn,
  Index,
  OneToMany,
  CreateDateColumn,
  UpdateDateColumn,
  OneToOne,
} from "typeorm";
import { VenueToilet } from "./VenueToilet.js";
import { VenueReview } from "./VenueReview.js";
import { VenueDashboardAnalytics } from "./VenueDashboardAnalytics.js";
import { VenueAccessibility } from "./VenueAccessibility.js";
import { VenueEnvironment } from "./VenueEnvironment.js";
import { VenueAdditionalInfo } from "./VenueAdditionalInformation.js";
import { Favorite } from "./Favourites.js";
import { Coordinates } from "./Coordinates.js";
import { OptionForAccess } from "./OptionForAccess.js";
import { AccessibilityStatus } from "./AccessibilityStatus.js";

@Entity("venue")
export class Venue extends BaseEntity {
  @PrimaryGeneratedColumn("uuid")
  venueId!: string;

  @Index()
  @Column({ name: "name", length: 100 })
  name!: string;

  @Column({ name: "phone", nullable: true })
  phone?: string;

  @Column({ name: "signature", nullable: true })
  signature?: string;

  @Index()
  @Column({ name: "email", nullable: true })
  email?: string;

  @Column({ type: "simple-array", name: "opening_hours", nullable: true })
  openingHours?: string[];

  @Column({ name: "front_image_of_business_describe", nullable: true })
  frontImageOfBusinessDescribe?: string;

  @Column({ name: "consent_obtained", default: false })
  consentObtained!: boolean;

  @Column({ name: "front_image", nullable: true })
  frontImage?: string;

  @Column({
    type: "simple-array",
    name: "accessible_parking_available",
    nullable: true,
  })
  accessibleParkingAvailable?: string[];

  @Column({ name: "closest_accessible_parking", nullable: true })
  closestAccessibleParking?: string;

  @Column({ name: "distance_to_closest_accessible_parking", nullable: true })
  distanceToClosestAccessibleParking?: string;

  // @Column({ type: "jsonb", nullable: true })
  // optionsForAccess: any;

  // New columns for venueInformation
  @Column({ name: "assessor_name", nullable: true })
  assessorName?: string;

  @Column({ name: "business_name", nullable: true })
  businessName?: string;

  @Column({ name: "business_type", nullable: true })
  businessType?: string;

  @Column({ type: "simple-array", nullable: true })
  address?: string[];

  @Column({ name: "front_image_of_business", nullable: true })
  frontImageOfBusiness?: string;

  @OneToOne(() => VenueToilet, (toilet) => toilet.venue)
  toilet!: VenueToilet;

  @OneToOne(() => AccessibilityStatus, (access) => access.venue)
  accessibilityStatus!: VenueToilet;

  @OneToMany(() => VenueReview, (review) => review.venue)
  review!: VenueReview[];

  @OneToMany(
    () => VenueDashboardAnalytics,
    (dashboardAnalytics) => dashboardAnalytics.venue
  )
  dashboardAnalytics!: VenueDashboardAnalytics[];

  // todo:: checking for now remove if not working
  @OneToMany(() => OptionForAccess, (option) => option.venue)
  optionsForAccess!: OptionForAccess[];

  @OneToOne(() => Coordinates, (coordinate) => coordinate.venue)
  coordinates!: Coordinates;

  @OneToOne(() => VenueAccessibility, (accessibility) => accessibility.venue)
  accessibility!: VenueAccessibility;

  @OneToOne(() => VenueEnvironment, (env) => env.venue)
  venueEnvironments!: VenueEnvironment;

  @OneToOne(() => VenueAdditionalInfo, (info) => info.venue)
  additionalInformation!: VenueAdditionalInfo;

  @OneToMany(() => Favorite, (favorite) => favorite.venue)
  favorites?: Favorite[];

  @CreateDateColumn({ name: "created_at", type: "time with time zone" })
  createdAt!: Date;

  @UpdateDateColumn({ name: "updated_at", type: "time with time zone" })
  updatedAt!: Date;
}
